### PR TITLE
Add crkbd keymap by manna harbour

### PR DESCRIPTION
**Pull request by manna-harbour**
This keymap includes crkbd-specific hardware feature support.  It provides easy selection of the following:

- hotswap trackpoint module in OLED port, or OLED with new or old drivers
- automatic mouse buttons layer activation on trackpoint movement
- keyboard-side mouse acceleration, or low rates for use with host-side acceleration
- static image display, or use as caps lock indicator
- built-in logo, or image from data file, with automatic rotation
- automatic conversion of image files
- matrix, light, or underglow RGB

It imports miryoku, an ergonomic, minimal, orthogonal layout for ergo or ortho keyboards.

## Description

This PR depends on the following open PRs:
- ~~#8805 "Add movement hook to ps2_mouse" to build with make argument MH_AUTO_BUTTONS=yes~~ [now merged]
- ~~#9442 "Update miryoku" to build with the default imported keymap~~ [now merged]

This PR would benefit from the following open PRs:
- ~~#9124 "Fix sharing of mouse button state from mousekeys to ps2_mouse" to eliminate extra mouse button clicks when building with make argument MH_AUTO_BUTTONS=yes~~ [now merged]

This PR provides solutions or work-arounds for the following closed issues:
- #1925 "[Question] Possibility to auto-switch layers when using trackpoint?" by building with make argument MH_AUTO_BUTTONS=yes
- #5982 "OLED doesn't turn off after timeout" with default macro MH_OLED_TIMEOUT
- #6594 "split keyboard halves in different state, can't reset" with default macro MH_RGB_TOGGLE_OFF

This PR provides enhancements to the following closed PRs:
- #4151 "Update logo and commonize it" by including a full size corne logo with make argument MH_OLED_IMAGE=corne.c
- #7237 "Added oled_write_raw and oled_write_raw_P functions to the OLED driver" by providing automatic conversion of image files with make argument MH_OLED_IMAGE=image.c
- #7535 "[keymap] Display logo for caps lock indication on crkbd using old OLED driver" by supporting the new OLED driver and arbitrary images with make arguments MH_MODULE=oled and MH_OLED_IMAGE=image.c
- #9142 "Add shift-to-disable to RGB_TOG keycode" by providing a keymap-level work around with default macro MH_RGB_TOGGLE_OFF

## Types of Changes

- [x] Keymap/layout/userspace (addition or update)

## Checklist

- [x] My code follows the code style of this project.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by manna-harbour** [comment]
As #9442 has been merged this will now build with the default configuration.  #8805 is still required to build with other options as described.

**Comment by manna-harbour** [comment]
All of the prerequisite PRs have now been merged and this is now ready for review.  Thanks!

**Comment by manna-harbour** [comment]
Thanks @drashna!

File: keyboards/crkbd/keymaps/manna-harbour/keymap.c [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+#include "ps2_mouse.h"
+#endif
+
+#if defined MOUSEKEY_ENABLE
+#include "mousekey.h"
+#endif
+
+#if defined SSD1306OLED
+#include "ssd1306.h"
+#endif
```
Comment (drashna) [inline keyboards/crkbd/keymaps/manna-harbour/keymap.c:23]: Ideally, the `SSD1306OLED` code shouldn't be used, at all. Instead, the oled driver feature should be used in its placen

**Comment by tzarc** [comment]
Can we also get GPL2+ compatible copyright headers on all source files, please?

File: keyboards/crkbd/keymaps/manna-harbour/config.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
+// https://github.com/manna-harbour/qmk_firmware/blob/crkbd/keyboards/crkbd/keymaps/manna-harbour/readme.org
```
Comment (drashna) [inline keyboards/crkbd/keymaps/manna-harbour/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/crkbd/keymaps/manna-harbour/config.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+#if defined OLED_DRIVER_ENABLE
+  #define OLED_TIMEOUT 30000 // ms
+  #define MH_OLED_TIMEOUT
+#endif
+
+#if defined RGB_MATRIX_ENABLE
+  #define RGB_MATRIX_KEYPRESSES
+  #define RGB_MATRIX_FRAMEBUFFER_EFFECTS
+  #define RGB_DISABLE_WHEN_USB_SUSPENDED true
```
Comment (drashna) [inline keyboards/crkbd/keymaps/manna-harbour/config.h:58]:
```suggestion
  #define RGB_DISABLE_WHEN_USB_SUSPENDED
```

File: keyboards/crkbd/keymaps/manna-harbour/config.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    #define RGBLED_NUM 6
+  #else
+    #define RGBLED_NUM 27
+  #endif
+  #define RGBLIGHT_SLEEP
+  #define RGBLIGHT_LIMIT_VAL 120
+  #define RGBLIGHT_HUE_STEP 8
+  #define RGBLIGHT_SAT_STEP 8
+  #define RGBLIGHT_VAL_STEP 8
+  #define RGBLIGHT_ANIMATIONS
```
Comment (drashna) [inline keyboards/crkbd/keymaps/manna-harbour/config.h:79]: Could you explicitly define which animations are enabled? 
```suggestion
#define RGBLIGHT_EFFECT_BREATHING
#define RGBLIGHT_EFFECT_RAINBOW_MOOD
#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#define RGBLIGHT_EFFECT_SNAKE
#define RGBLIGHT_EFFECT_KNIGHT
#define RGBLIGHT_EFFECT_CHRISTMAS
#define RGBLIGHT_EFFECT_STATIC_GRADIENT
#define RGBLIGHT_EFFECT_RGB_TEST
#define RGBLIGHT_EFFECT_ALTERNATING
#define RGBLIGHT_EFFECT_TWINKLE
```

File: keyboards/crkbd/keymaps/manna-harbour/keymap.c [hunk 5]
```diff
@@ -0,0 +1,181 @@
+// https://github.com/manna-harbour/qmk_firmware/blob/crkbd/keyboards/crkbd/keymaps/manna-harbour/readme.org
```
Comment (drashna) [inline keyboards/crkbd/keymaps/manna-harbour/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/crkbd/keymaps/manna-harbour/keymap.c (hunk 1 above)
Comment (drashna) [reply keyboards/crkbd/keymaps/manna-harbour/keymap.c:23]:
```suggestion
```

File: keyboards/crkbd/keymaps/manna-harbour/keymap.c [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+void mh_draw_image(void) {
+  oled_write_raw_P(mh_image, mh_image_size);
+}
+
+#endif // defined MH_OLED_IMAGE_FILE && defined OLED_DRIVER_ENABLE
+
+
+#if defined OLED_DRIVER_ENABLE
```
Comment (drashna) [inline keyboards/crkbd/keymaps/manna-harbour/keymap.c:91]:
```suggestion
#if defined OLED_ENABLE
```

File: keyboards/crkbd/keymaps/manna-harbour/keymap.c [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+  matrix_clear(&matrix);
+  matrix_render_user(&matrix);
+  matrix_update(&display, &matrix);
+}
+
+void matrix_scan_user(void) {
+   iota_gfx_task();
+}
+
+#endif // defined SSD1306OLED
```
Comment (drashna) [inline keyboards/crkbd/keymaps/manna-harbour/keymap.c:173]:
```suggestion
```

File: keyboards/crkbd/keymaps/manna-harbour/rules.mk [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+## rules.mk, config.h, manna-harbour_miryoku.h, manna-harbour_miryoku.c
+USER_NAME := manna-harbour_miryoku
+
+
+this_dir=keyboards/crkbd/keymaps/manna-harbour
+
+OPT_DEFS += -DMH_USER_NAME_H=\"$(USER_NAME).h\"
+
+ifeq ($(strip $(MH_MODULE)), trackpoint)
```
Comment (drashna) [inline keyboards/crkbd/keymaps/manna-harbour/rules.mk:42]: All of this should probably go in `post_rules.mk`

File: keyboards/crkbd/keymaps/manna-harbour/mh_images/converted/corne.c [hunk 9]
```diff
@@ -0,0 +1,40 @@
+// https://github.com/manna-harbour/qmk_firmware/blob/crkbd/keyboards/crkbd/keymaps/manna-harbour/readme.org
```
Comment (drashna) [inline keyboards/crkbd/keymaps/manna-harbour/mh_images/converted/corne.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

File: keyboards/crkbd/keymaps/manna-harbour/mh_images/converted/qmk.c [hunk 10]
```diff
@@ -0,0 +1,40 @@
+// https://github.com/manna-harbour/qmk_firmware/blob/crkbd/keyboards/crkbd/keymaps/manna-harbour/readme.org
```
Comment (drashna) [inline keyboards/crkbd/keymaps/manna-harbour/mh_images/converted/qmk.c:1]: Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by keyboard-magpie** [comment]
@manna-harbour in the spirit of tidying up, does this PR still need to remain open, or does your current mechanism fulfil all needs? IMO the latter is true.

**Comment by manna-harbour** [comment]
Thanks @drashna and @tzarc for the review. 

This PR is now very out of date, and I've found a different method of adding custom code to any keymap at https://github.com/manna-harbour/qmk_firmware/issues/40, which should be much better than the `USER_NAME` method used here. The features from this PR will be added there in future, and I'll be sure to incorporate your feedback from this PR at that time.

As an aside, there was 12 months between @drashna's original approval and request for review, and the next review. There may be an issue with the use of the `awaiting review` label, as it suppresses the stale action, which seems to be the only automated reminder for PRs needing attention. Perhaps an `awaiting review` reminder action is also needed?

Thanks @keyboard-magpie for the reminder. The Miryoku [customisation](https://github.com/manna-harbour/miryoku/discussions/85) method is separate from this, but even so, this PR is now obsolete. Closing now.

**Pull request closed without merging**